File: rvb-sim/dnp3/Helpers/AnalogOutputControl.vb
Imports System.Threading.Tasks
Imports Automatak.DNP3.Interface

Public Class AnalogOutputControl
    Enum AOType
        Integer16
        Integer32
        Float32
        Double64
    End Enum

    Public ReadOnly Property Type As AOType
    Public ReadOnly Property Value As Integer
    Public ReadOnly Property Index As Integer

    Public Sub New(ByVal type As AOType, ByVal value As Integer, ByVal index As Integer)
        Me.Type = type
        Me.Value = value
        Me.Index = index
    End Sub

    Public ReadOnly Property DirectOperateAction As Func(Of ICommandProcessor, Task(Of CommandTaskResult))
        Get
            Dim index = Me.Index
            Dim value = Me.Value
            ' Dim type = Me.Type

            Select Case Type
                Case AOType.Integer16
                    Return Function(ByVal cp) cp.DirectOperate(New AnalogOutputInt16(Convert.ToInt16(value)), index, TaskConfig.[Default])
                Case AOType.Integer32
                    Return Function(ByVal cp) cp.DirectOperate(New AnalogOutputInt32(Convert.ToInt32(value)), index, TaskConfig.[Default])
                Case AOType.Float32
                    Return Function(ByVal cp) cp.DirectOperate(New AnalogOutputFloat32(Convert.ToSingle(value)), index, TaskConfig.[Default])
                Case Else
                    Return Function(ByVal cp) cp.DirectOperate(New AnalogOutputDouble64(Convert.ToDouble(value)), index, TaskConfig.[Default])
            End Select
        End Get
    End Property

    Public ReadOnly Property SelectAndOperateAction As Func(Of ICommandProcessor, Task(Of CommandTaskResult))
        Get
            Dim index = Me.Index
            Dim value = Me.Value

            Select Case Type
                Case AOType.Integer16
                    Return Function(ByVal cp) cp.SelectAndOperate(New AnalogOutputInt16(Convert.ToInt16(value)), index, TaskConfig.[Default])
                Case AOType.Integer32
                    Return Function(ByVal cp) cp.SelectAndOperate(New AnalogOutputInt32(Convert.ToInt32(value)), index, TaskConfig.[Default])
                Case AOType.Float32
                    Return Function(ByVal cp) cp.SelectAndOperate(New AnalogOutputFloat32(Convert.ToSingle(value)), index, TaskConfig.[Default])
                Case Else
                    Return Function(ByVal cp) cp.SelectAndOperate(New AnalogOutputDouble64(Convert.ToDouble(value)), index, TaskConfig.[Default])
            End Select
        End Get
    End Property
End Class

File: rvbsim.dnp/Helpers/AnalogOutputControl.cs
using System;
using System.Threading.Tasks;
using Automatak.DNP3.Interface;

namespace rvb_sim.dnp
{
    public class AnalogOutputControl
    {

        public ushort Index { get; }
        public decimal Value { get; }
        public AOType Type { get; }

        public enum AOType
        {
            Integer16,
            Integer32,
            Float32,
            Double64
        };

        public AnalogOutputControl() { }

        public AnalogOutputControl(ushort index, decimal value, AOType type)
        {
            Index = index;
            Value = value;
            Type = type;
        }

        //public void DirectOperateAction()
        //{

        //}

        public Func<ICommandProcessor, Task<CommandTaskResult>> DirectOperateAction
        {
            get
            {
                var index = Index;
                var value = Value;
                var type = Type;

                switch (type)
                {
                    case (AOType.Integer16):
                        return (ICommandProcessor cp) => cp.DirectOperate(new AnalogOutputInt16(SafeConvert.ToInt16(value)), index, TaskConfig.Default);
                    case (AOType.Integer32):
                        return (ICommandProcessor cp) => cp.DirectOperate(new AnalogOutputInt32(SafeConvert.ToInt32(value)), index, TaskConfig.Default);
                    case (AOType.Float32):
                        return (ICommandProcessor cp) => cp.DirectOperate(new AnalogOutputFloat32(SafeConvert.ToSingle(value)), index, TaskConfig.Default);
                    default:
                        return (ICommandProcessor cp) => cp.DirectOperate(new AnalogOutputDouble64(SafeConvert.ToDouble(value)), index, TaskConfig.Default);
                }
            }
        }

        public Func<ICommandProcessor, Task<CommandTaskResult>> SelectAndOperateAction
        {
            get
            {

                var index = Index;
                var value = Value;
                var type = Type;

                switch (Type)
                {
                    case (AOType.Integer16):
                        return (ICommandProcessor cp) => cp.SelectAndOperate(new AnalogOutputInt16(Convert.ToInt16(Value)), index, TaskConfig.Default);
                    case (AOType.Integer32):
                        return (ICommandProcessor cp) => cp.SelectAndOperate(new AnalogOutputInt32(Convert.ToInt32(value)), index, TaskConfig.Default);
                    case (AOType.Float32):
                        return (ICommandProcessor cp) => cp.SelectAndOperate(new AnalogOutputFloat32(Convert.ToSingle(value)), index, TaskConfig.Default);
                    default:
                        return (ICommandProcessor cp) => cp.SelectAndOperate(new AnalogOutputDouble64(Convert.ToDouble(value)), index, TaskConfig.Default);
                }
            }
        }
    }
}
